### PR TITLE
Kyle/stream resources

### PR DESCRIPTION
**Pull request by byubean**

### CODE REVIEW

File: src/quest/historian.py [hunk 1]
```diff
@@ -970,33 +971,48 @@ async def stream_resources(self, identity):
 
         logging.debug(f'Resource stream created for {identity}')
 
+        stream_id = uuid.uuid4()
```
Comment (byubean) [inline src/quest/historian.py:974]: I think we want the `stream_id` in the log line just above.

File: src/quest/historian.py [hunk 2]
```diff
@@ -970,33 +971,48 @@ async def stream_resources(self, identity):
...
 
         while not self._workflow_completed:
             # Yield new resources updates as they become available
+            self._resource_events[identity]['stream_request'].set()
             await self._resource_queues[identity].get()
+            self._resource_events[identity]['stream_gate'].set()
             yield await self.get_resources(identity)
-            self._resource_events[identity].set()
+
+    async def close_resource_stream(self, identity):
```
Comment (byubean) [inline src/quest/historian.py:997]: Don't we want to close the resource stream by the `stream_id`?

File: src/quest/historian.py [hunk 3]
```diff
@@ -970,33 +971,48 @@ async def stream_resources(self, identity):
...
             assert self._resource_events.get(identity) is not None
 
-            self._resource_events[identity].clear()
-            await resource_queue.put('Resource Update')
-            await self._resource_events[identity].wait()
+            if self._resource_events[identity]['stream_request'].is_set():
+                self._resource_events[identity]['stream_request'].clear()
+                await resource_queue.put('Resource Update')
+                await self._resource_events[identity]['stream_gate'].wait()
+                self._resource_events[identity]['stream_gate'].clear()
```
Comment (byubean) [inline src/quest/historian.py:1015]: Let's add comments to this also.

File: src/quest/historian.py [hunk 4]
```diff
@@ -970,33 +971,48 @@ async def stream_resources(self, identity):
...
         if (resource_queue := self._resource_queues.get(identity)) is not None:
             # If the user has called stream_resources, there should be a resource_event for `identity`
             assert self._resource_events.get(identity) is not None
 
-            self._resource_events[identity].clear()
-            await resource_queue.put('Resource Update')
-            await self._resource_events[identity].wait()
+            if self._resource_events[identity]['stream_request'].is_set():
+                self._resource_events[identity]['stream_request'].clear()
+                await resource_queue.put('Resource Update')
```
Comment (byubean) [inline src/quest/historian.py:1013]: I'd like to use an `asyncio.Event` here instead of a queue with dummy values.

**Comment by byubean** [comment]
Here's my (evolving, searching) thinking on this feature:

The task of pushing a notification for a resource update is necessarily a **workflow** task. The workflow knows that a resource has changed, and therefore the workflow must notify the listeners. 

However, the workflow and the listener are running on separate tasks. Because they are on a separate tasks, the workflow must await the outside task in order to give that task the chance to respond to the resource change. Then the outside task must await the workflow task in order to give the workflow right-of-way to continue. 

We are temporarily allowing (forcing) the workflow's train-of-thought to cross over to another task in order to make the notification before returning to the workflow task. Because of python's single-threaded nature, the only way for the notifications to happen is for the workflow to do it. 

To add a listener to resource changes is to modify the workflow. The workflow must now do more than it did before. This is probably a good and useful feature, but those adding listeners should be educated and responsible enough to not violate the trust. Don't add listeners that are going to crash.

So, how do we handle socket failures? If I have a task streaming resources from a workflow, and I try to send data on a socket in response to resource changes, and then it throws an error -- then what?

I think the workflow should not care about the listeners (or whether there are any listeners). It should care about its resources. For example, the workflow is going to set state and listen on queues. The workflow doesn't know or care how the outside world reads that data nor how the outside world puts data on the queues. 

So, while the workflow is required to notify anyone that is listening for resource updates, as soon as the listener leaves, the workflow stops caring. Thus, we need to be sure that if there is an exception in the listener, the connection is immediately closed. This gets us to the `ResourceStream` context design.

So, because the listeners are always on a separate task than the workflow, I can have multiple listeners operating in parallel. This is good.

Adding a ResourceStream context that opens and closes a resource stream will ensure that if an error happens in the listener, the stream closes and the workflow moves on. We'll need to handle the situation where the workflow is waiting on a listener gate, the listener leaves, the stream closes, and the gate is deleted -- but is currently being awaited? I think that when we close a stream, we set all the gates, and **then** we delete them from the dictionaries. 

In conclusion, I think we actually do need the coupled design between workflow and listener. But we do need the ability for a listener to leave and decouple the workflow.

File: demos/multi_guess_client.py [hunk 5]
```diff
@@ -52,7 +52,7 @@ async def main():
         # until they are no longer interested.
 
         # What I want:
-        async with client.get_resource_stream('demo', None) as resource_stream:
+        async with client.stream_resources('demo', None) as resource_stream:
```
Comment (byubean) [inline demos/multi_guess_client.py:55]: Now that we have the no-with-needed design, we can get rid of this "What I want" section and correct the comments below.

File: quest_test/test_resource_stream.py [hunk 6]
```diff
@@ -1,11 +1,19 @@
+import uuid
+
 import pytest
+from pathlib import Path
 
-from src.quest import step, Historian
+from src.quest import step, create_filesystem_manager
```
Comment (byubean) [inline quest_test/test_resource_stream.py:6]: For tests we can use the historian directly. Or at least use an in-memory manager.

File: quest_test/test_resource_stream.py [hunk 7]
```diff
@@ -14,60 +22,76 @@ async def workflow():
         async with state('phrase', None, 'woot') as phrase:
             await phrase.set(await big_phrase(await phrase.get()))
 
-            async with queue('messages', None) as messages:
-                new_message = await messages.get()
-                await phrase.set((await phrase.get()) + new_message)
+            # async with queue('messages', None) as messages:
+            #     new_message = await messages.get()
+            #     await phrase.set((await phrase.get()) + new_message)
```
Comment (byubean) [inline quest_test/test_resource_stream.py:27]: Why are these commented out?

File: quest_test/test_resource_stream.py [hunk 8]
```diff
@@ -14,60 +22,76 @@ async def workflow():
...
-    except StopAsyncIteration:
-        pass
-
-    await wtask
+    w_id = 'input-' + str(uuid.uuid1())
+    async with create_filesystem_manager(
+            Path('test_state'),
+            'resource_stream_test',
+            lambda workflow_id: workflow
+    ) as manager:
```
Comment (byubean) [inline quest_test/test_resource_stream.py:36]: using an in-memory manager avoids needing to hit the filesystem in a unit test (which is preferred).

File: quest_test/test_resource_stream.py [hunk 9]
```diff
@@ -14,60 +22,76 @@ async def workflow():
...
+        manager.start_workflow('', w_id)
+        # TODO: Discuss whether stream_resources should be async or not. (And make so in manager, historian, resources)
```
Comment (byubean) [inline quest_test/test_resource_stream.py:38]: If it doesn't need to be async, it shouldn't be.

File: quest_test/test_resource_stream.py [hunk 10]
```diff
@@ -14,60 +22,76 @@ async def workflow():
...
+        index = 0
+        async for resources in manager.stream_resources(w_id, None):
+            if index == 0:
+                manager.start_workflow('', w_id)
+            print(f'Resources: {resources}')
+            index += 1
+
+        # This isn't implemented in manager...
+        # await manager.wait_for_completion(w_id, None)
```
Comment (byubean) [inline quest_test/test_resource_stream.py:47]: We don't want this anymore. Now that the stream closes when the loop exits, we should be able to `break` from the loop and be ok.

That's a test we should try. We know that throwing an exception closes a context in the generator function. What if the client code `break`s from the loop? What if I `return` from within a loop? Does the context in the generator still get closed?

File: quest_test/test_resource_stream.py [hunk 11]
```diff
@@ -14,60 +22,76 @@ async def workflow():
...
+
+# This is from old version of test stream resources
```
Comment (byubean) [inline quest_test/test_resource_stream.py:49]: I think the code below should still work for our "vanilla" test.

File: scratch/async_generator_scratch.py [hunk 12]
```diff
@@ -1,37 +1,43 @@
...
+import random
 
-my_queue = asyncio.Queue()
 
-async def fetch_data():
-    input_num = input('Please enter your number: ')
-    return f'The given number is: {input_num}'
+class StreamContextManager:
+    def __init__(self):
+        pass
```
Comment (byubean) [inline scratch/async_generator_scratch.py:7]: You shouldn't need this code.

File: src/quest/historian.py [hunk 13]
```diff
@@ -2,6 +2,7 @@
 import inspect
 import logging
 import traceback
+import uuid
```
Comment (byubean) [inline src/quest/historian.py:5]: Is this import still used?

File: src/quest/historian.py [hunk 14]
```diff
@@ -955,55 +951,15 @@ async def get_resources(self, identity):
...
+            lambda: self._workflow_completed
+        )
 
-    async def wait_for_completion(self, identity):
-        """
-        Wait for the workflow to complete after starting the resource stream
-        """
-        async for _ in self.stream_resources(identity):
-            pass
+    async def update_resource_stream(self, identity):
```
Comment (byubean) [inline src/quest/historian.py:961]: This should be `_update_resource_stream` (private visibility)

File: src/quest/resources.py [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+
+class ResourceStreamManager:
+    def __init__(self):
+        self.resource_streams: dict[str, set[ResourceStreamManager.ResourceStream]] = {}
+
+    class ResourceStreamContext:
+        def __init__(self, on_close: Callable):
+            self._on_close = on_close
+
+        # TODO: Neither of these seem to actually need to be async? Can we make them normal? Problems with doing async things under and normal with context?
```
Comment (byubean) [inline src/quest/resources.py:14]: If we don't actually need an async context, we should use a normal one.

Comment (byubean) [reply src/quest/resources.py:14]: Does the generator-context-close phenomenon work when the context is not async?

File: src/quest/resources.py [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+                     get_resources: Callable[[], Coroutine],
+                     is_workflow_complete: Callable[[], bool],
+                     on_close: Callable[['ResourceStreamManager.ResourceStream'], None]
+                     ):
+            self._stream_gate = asyncio.Event()
+            self._update_event = asyncio.Event()
+            self._get_resources = get_resources
+            self._is_workflow_complete = is_workflow_complete
+            self._on_close = on_close
+            # TODO - set any gates?
```
Comment (byubean) [inline src/quest/resources.py:34]: Thoughts on this? Are the default settings the ones we need?

File: src/quest/resources.py [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+        self.resource_streams[identity].add(rs)
+        return rs
+
+    async def update(self, identity):
+        # If there is no resources stream associated with `identity`, no update needed.
+        if identity not in self.resource_streams:
+            return
+
+        # TODO: Say a stream is added but then removed, the identity still exists as a key but just has an empty set.
+        # Is it okay to leave it as such?
```
Comment (byubean) [inline src/quest/resources.py:81]: I think we should remove empty sets as we go. Otherwise we get a memory leak as we keep track of more and more empty sets over time.

**Comment by kjstanding** [comment]
Closed PR to clean up comments for new version of this branch
